### PR TITLE
Fix Collection selection using SwipeView

### LINKED ISSUE #8774: [Bug] SwipeView in CollectionView selection
Using SwipeView in CollectionView: SwipeView catch touch gesture and not propagate it to CollectionView so SelectionChanged is no invoked

When i try to evade with TapGestureRecognizer on SwipeView, then SwipeItem Command and TapGesture Command are invoked at same time. (Android only)

### Steps to Reproduce

1. Put SwipeView (with some SwipeItems) in CollectionView
2. Set CollectionView SelectionMode to Single and attach event SelectionChanged
3. After touching item, event SelectionChanged is not invoked

### Expected Behavior
On touch only on SwipeView, SwipeView should propagate touch event to CollectionView (+ Commands, ProperyBindings, ....)

### Example

[CollectionViewAndSwipeView.zip](https://github.com/xamarin/Xamarin.Forms/files/3930537/CollectionViewAndSwipeView.zip)

### Basic Information
- Version with issue: XF 4.4.0.991220-pre3
- IDE: VS 2019 16.4
- Platform Target Frameworks:
  - iOS:  Not working (iOS 13.2.3)
  - Android: Not working (SDK Android 9.0)
  - UWP:  On UWP SwipeView not works in CollectionView 
- Android Support Library Version: 28.0.0.3

### LINKED ISSUE #8903: [Enhancement] Use SwipeView with VisualState Selection in CollectionView
## Summary
Can we use VisualState Selected and SwipeView in CollectionView ?

``` 
...
<ContentPage.Resources>
        <Style TargetType="Frame">
            <Setter Property="VisualStateManager.VisualStateGroups">
                <VisualStateGroupList>
                    <VisualStateGroup x:Name="CommonStates">
                        <VisualState x:Name="Normal" />
                        <VisualState x:Name="Selected">
                            <VisualState.Setters>
                                <Setter Property="BackgroundColor" Value="LightSkyBlue" />
                            </VisualState.Setters>
                        </VisualState>
                    </VisualStateGroup>
                </VisualStateGroupList>
            </Setter>
        </Style>
</ContentPage.Resources>
...
```
as the root of DataTemplate is SwipeView, the Style TargetType is set to **SwipeView**, the selectedItem appear unselected, even we set the Style TargetType to **Frame** always not working
```
... 

<CollectionView Grid.Row="0" ItemsSource="{Binding Items}" SelectionMode="Single">
	<CollectionView.ItemsLayout>
		<LinearItemsLayout Orientation="Vertical" ItemSpacing="3" />
	</CollectionView.ItemsLayout>
	<CollectionView.ItemTemplate>
		<DataTemplate x:DataType="models:MyModel">
			<SwipeView>
				<SwipeView.LeftItems>
					<SwipeItems>
						<SwipeItem Text="Delete" BackgroundColor="Red"/>
					</SwipeItems>
				</SwipeView.LeftItems>
				<SwipeView.RightItems>
					<SwipeItems>
						<SwipeItem Text="Synch." BackgroundColor="DarkGreen"/>
					</SwipeItems>
				</SwipeView.RightItems>

				<Frame BorderColor="LightGray">
					<Grid Padding="10">
							<Label Text="{Binding Label}" />
						</Grid>
					</Grid>
				</Frame>
			</SwipeView>
		</DataTemplate>
	</CollectionView.ItemTemplate>
</CollectionView>
...	
```
can anyone give us a help, thanks in advance

### LINKED ISSUE #9027: [Bug] CollectionView Selection Changed (or item tapped) not working using SwipeView on Android
### Description
If I use SwipeView in a CollectionView ItemTemplate, then, SelectionChanged and Item Tapped is not working.
### Steps to Reproduce
Having the following XAML  
  ```
<CollectionView ItemsSource="{Binding Items}"
                        Margin="0,16,0,0"
                        Grid.Row="1"
                        SelectionMode="Single">
            <CollectionView.ItemTemplate>
                <DataTemplate>
                    <Grid Margin="6">
                        <SwipeView>
                            <SwipeView.LeftItems>
                                <SwipeItems Mode="Execute">
                                    <SwipeItemView BackgroundColor="White">
                                        <Label TextColor="Blue"
                                               VerticalOptions="Center"
                                               Text="Test1"
                                               VerticalTextAlignment="Center"
                                               Padding="10" />

                                    </SwipeItemView>
                                </SwipeItems>
                            </SwipeView.LeftItems>
                            <SwipeView.RightItems>
                                <SwipeItems Mode="Execute">
                                    <SwipeItemView BackgroundColor="White"
                                                   VerticalOptions="Center">
                                        <Label TextColor="Red"
                                               VerticalOptions="Center"
                                               Text="Test2"
                                               VerticalTextAlignment="Center"
                                               Padding="10" />
                                    </SwipeItemView>
                                </SwipeItems>
                            </SwipeView.RightItems>
                            <Frame Padding="4"
                                   Margin="0"
                                   BackgroundColor="LightGray">
                                <Frame.GestureRecognizers>
                                    <TapGestureRecognizer Tapped="TapGestureRecognizer_Tapped_1" />
                                </Frame.GestureRecognizers>

                                <Label TextColor="Black"
                                       FontSize="24"
                                       FontAttributes="Bold"
                                       Text="I'm an Item" />
                            </Frame>
                        </SwipeView>
                    </Grid>
                </DataTemplate>
            </CollectionView.ItemTemplate>
        </CollectionView>
```
TapGestureRecognizer_Tapped_1 never gets called, neither if you add a SelectionChanged event in CollectionView
### Expected Behavior
tap event should get called, selectionChanged event should get called
### Actual Behavior
TapGestureRecognizer_Tapped_1 never gets called, neither if you add a SelectionChanged event in CollectionView
### Basic Information

- Version with issue: 4.4.0.991265
- Last known good version:
- IDE: Visual Studio Community 2017  Version 15.9.17
- Platform Target Frameworks: 

  - Android: 8.1

### LINKED ISSUE #9089: Proper support for SwipeView in ListView with Tap Command
I have a ListView with each cell containing a SwipeView. In order to support a Tap command on each cell to open the item, I've added a GestureRecognizer, which seems to work. However, I only want the tap to open the item when the SwipeView's context menu is closed. When the context menu is opened, a tap should just close the context menu.  

Seemingly, the way to do this would be to have a boolean IsOpened that I can check in the OnTapped event, but SwipeView does not have support for this. I've tried to extend the SwipeView using the provided SwipeEnded event which is fired when the context menu is opened, but the CloseRequested event is only triggered when the Close() function is programmatically called, so there's no way to know when the context has been closed.

Alternatively, I would look for the Context view's offset or something that might show that it's open, but I can't seem to determine which parameters to utilize. 

I'm not sure if this is a bug or intended to work this way, but CloseRequested should fire when the context menu is closed programmatically or automatically. Further, having some boolean such as IsOpen to determine if the context menu is opened/closed would be helpful.

```xml
<ListView
    ItemsSource="{Binding Items}"
    SelectionMode="None"
    CachingStrategy="RecycleElement"
    RowHeight="110"
    HeightRequest="1000"
    BackgroundColor="White"
    IsPullToRefreshEnabled="True"
    Refreshing="OnRefresh">
    <ListView.ItemTemplate>
        <DataTemplate>
            <ViewCell>
                <SwipeView>                    
                    <SwipeView.GestureRecognizers>
                        <TapGestureRecognizer
                            Tapped="OnTapped"
                            NumberOfTapsRequired="1" />
                    </SwipeView.GestureRecognizers>                    
                    <SwipeView.RightItems>
                        <SwipeItems>
                            <SwipeItem
                                Text="Edit"
                                IconImageSource="edit_icon.png"
                                BackgroundColor="Green"
                                Invoked="OnEdit" />
                            <SwipeItem
                                Text="Delete"
                                IconImageSource="delete_icon.png"
                                BackgroundColor="Red"
                                Invoked="OnDelete" />
                        </SwipeItems>
                    </SwipeView.RightItems>
                    <SwipeView.Content>
                        :
                    </SwipeView.Content>
                </SwipeView>
            </ViewCell>
        </DataTemplate>
    </ListView.ItemTemplate>
</ListView>
```

### LINKED ISSUE #9466: [Bug] Selection event not working fine when we use swipeview with Listiew
### Description
I tried the swipeview sample,   Swipe from left to right and right to left working fine. But when i try to execute the command parameter or selection of that one item in list sele
Swipeview issue
ction is not working . For instance am attaching the screen shot about the issue, below screen shot you can find the 3 dots symbol and small plus icon so those click are not working fine after add swipe view. Please let me know. Thanks.

### Steps to Reproduce

1. Try to add swipeview with Listview
2. Try to add buttons in item in listview
3. Perform command parameter 

### Expected Behavior
It should be work click event also
### Actual Behavior
Command parameters not working fine 
### Basic Information

- Version with issue: 4.4 Xamarin forms
- Last known good version: 4.3
- IDE:Visual studio 2019
- Platform Target Frameworks: 
  - iOS:  Working fine
  - Android: Not working fine
  -
- Android Support Library Version: 9

- Affected Devices: Oppo

### Screenshots :- Attached
Swipeview issue

### LINKED ISSUE #9467: ListView itemTapped Event Not fired if view cell contains Swipe view
Using xamarin forms v 4.4.0.991265 & visual studio 16.4.

On using swipe view inside view cell of list view the item tapped event of list view is disabled.

```       <ListView.Behaviors>
                    <ETC:EventToCommandBehavior EventName="ItemTapped"
                     Command="{Binding Source={x:Reference Name=root}, 
                   Path=BindingContext.ListViewItemTappedCommand}" />
          </ListView.Behaviors>

         <ListView.ItemTemplate>
                    <DataTemplate>
                        <ViewCell Height="100" >
                            <ViewCell.View>
                                <StackLayout HeightRequest="100">
                                    <SwipeView>
                                        <SwipeView.LeftItems>
                                            <SwipeItems Mode="Reveal"
                                               SwipeBehaviorOnInvoked="RemainOpen">
                                                <SwipeItemView >
                                                    <Grid WidthRequest="100">
                                                        <Frame BackgroundColor="DarkGray" CornerRadius="30">
                                                            <Label Text="1"
                                                                 HorizontalOptions="Center"
                                                                VerticalOptions="Center" />
                                                        </Frame>
                                                    </Grid>
                                                </SwipeItemView>
                                            </SwipeItems>
                                        </SwipeView.LeftItems>
                                        <Grid>
                                            <Grid.ColumnDefinitions>
                                                <ColumnDefinition Width="*"/>
                                                <ColumnDefinition Width="3*"/>
                                                <ColumnDefinition Width="auto"/>
                                                <ColumnDefinition Width="*"/>
                                            </Grid.ColumnDefinitions>

                                            <Image Source="UserAvatar" Grid.Column="0"/>
                                            <StackLayout Grid.Column="1"
                                             HorizontalOptions="Center" 
                                             VerticalOptions="Center">

                                                <Label Text="{Binding VisitorName}" />
                                                <Label Text="{Binding VisitorPhoneNumber}" />
                                                <Label>
                                                    <Label.FormattedText>
                                                        <FormattedString>
                                                            <FormattedString.Spans>
                                                                <Span Text="{Binding Visitdate}"/>
                                                                <Span Text="{Binding VisitDuration}"/>
                                                            </FormattedString.Spans>
                                                        </FormattedString>
                                                    </Label.FormattedText>
                                                </Label>
                                            </StackLayout >
                                            <BoxView VerticalOptions="FillAndExpand" WidthRequest="1" 
                                             Grid.Column="2" BackgroundColor="{DynamicResource ColorFive}" 
                                             Margin="0,15,0,15"/>
                                            <StackLayout Grid.Column="3" VerticalOptions="Center">
                                                <Label Text="{Binding status}" 
                                                TextColor="{DynamicResource  MaroonMainColor}" />
                                                <Label Text="{Binding Response}" 
                                                 TextColor="{DynamicResource MaroonMainColor}"/>
                                            </StackLayout>
                                        </Grid>
                                    </SwipeView>
                                </StackLayout>
                            </ViewCell.View>
                        </ViewCell>
                    </DataTemplate>
                </ListView.ItemTemplate>
            </ListView>
```

### LINKED ISSUE #10056: [Bug] [iOS] SwipeView items don't support TapGestureRecognizer
### Description
When using SwipeView on Android and adding a TapGestureRecognizer to a SwipeView item, swiping and tapping behaves as expected. On iOS the taps are ignored.

### Steps to Reproduce

1. Add a swipeview item to your view
2. Add a TapGestureRecognizer
3. Tap on iOS - nothing happens
4. Tap on Android - event handler is fired

### Expected Behavior
TapGestureRecognizer event handler is fired

### Actual Behavior
Nothing happens

### Basic Information

- Version with issue: 4.5.0.396
- Last known good version: NA
- IDE: Visual Studio for Windows 16.4.5, Visual Studio for Mac 8.4.8 build 2
- Platform Target Frameworks: 
  - iOS:  12.2 and 13.3 both tested
  - Android: 28

### Reproduction Link

Working on this, but sample code:

```
<SwipeView>
         <SwipeView.GestureRecognizers>
                <TapGestureRecognizer NumberOfTapsRequired="1" 
                                                      Tapped="TapGestureRecognizer_Tapped"/>
         </SwipeView.GestureRecognizers>
                <SwipeView.LeftItems>
                    <SwipeItems>
                        <SwipeItem Text="Example swipe item"
                                   BackgroundColor="Pink"
                                   />
                    </SwipeItems>
                </SwipeView.LeftItems>
                <StackLayout>
                       <Label Text="Example"/>
               </StackLayout>
  </SwipeView>
```

### Workaround

None

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change ###

Fix Collection selection (CollectionView and ListView) using SwipeView.

Fixed on Android and iOS. 

### Issues Resolved ### 

- fixes #9467
- fixes #8774
- fixes #8903
- fixes #9089
- fixes #9466
- fixes #9027
- fixes #10056

### API Changes ###

 None

### Platforms Affected ### 

- iOS
- Android

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Android
fix9466

iOS
fix9467-ios

Using GestureRecognizers
77737025-764f0f80-700d-11ea-8902-91240ddf635f

### Testing Procedure ###
Launch Core Gallery and navigate to the SwipeView samples. Select the ListView and/or the CollectionView sample. Then, tap any item and verify that the SelectionChanged event of the CollectionView is executed.

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by mdcsoln** [comment]
Hi All,

Could you let us know, when can be this changes will be available in next release?

Thanks,
Suhas

**Comment by lhughey** [comment]
I hope this PR can be reviewed and merged. This functionality is greatly needed.

**Comment by jayrlsw** [comment]
> I hope this PR can be reviewed and merged. This functionality is greatly needed.

Yes please!

**Comment by arunr14** [comment]
Is there a target date or release to merge this in ?

File: Xamarin.Forms.Platform.Android/Cells/ViewCellRenderer.cs [hunk 1]
```diff
@@ -141,12 +161,18 @@ public override bool DispatchTouchEvent(MotionEvent e)
 
 				if (_watchForLongPress)
 				{
-					// Feed the gestue through the LongPress detector; for this to wor we *must* return true 
+					// Feed the gesture through the LongPress detector; for this to wor we *must* return true
```
Comment (jfversluis) [inline Xamarin.Forms.Platform.Android/Cells/ViewCellRenderer.cs]:
```suggestion
					// Feed the gesture through the LongPress detector; for this to work we *must* return true 
```

File: Xamarin.Forms.Controls/GalleryPages/SwipeViewGalleries/SwipeBindableLayoutGallery.xaml.cs [hunk 2]
```diff
@@ -49,7 +49,7 @@ public ObservableCollection<Message> Messages
 
 		public ICommand FavouriteCommand => new Command(OnFavourite);
 		public ICommand DeleteCommand => new Command(OnDelete);
-
+		public ICommand TapCommand => new Command(OnTap);
 
 		private void LoadMessages()
```
Comment (jfversluis) [inline Xamarin.Forms.Controls/GalleryPages/SwipeViewGalleries/SwipeBindableLayoutGallery.xaml.cs]:
```suggestion
		void LoadMessages()
```

File: Xamarin.Forms.Controls/GalleryPages/SwipeViewGalleries/SwipeListViewGallery.xaml.cs [hunk 3]
```diff
@@ -13,5 +13,10 @@ public SwipeListViewGallery()
 			MessagingCenter.Subscribe<SwipeViewGalleryViewModel>(this, "favourite", sender => { DisplayAlert("SwipeView", "Favourite", "Ok"); });
 			MessagingCenter.Subscribe<SwipeViewGalleryViewModel>(this, "delete", sender => { DisplayAlert("SwipeView", "Delete", "Ok"); });
 		}
+
+		async void OnSwipeListViewItemTapped(object sender, ItemTappedEventArgs args)
+		{
+			await DisplayAlert("OnSwipeListViewItemTapped", "You have been tapped a ListView item", "Ok");
```
Comment (jfversluis) [inline Xamarin.Forms.Controls/GalleryPages/SwipeViewGalleries/SwipeListViewGallery.xaml.cs]:
```suggestion
			await DisplayAlert("OnSwipeListViewItemTapped", "You have tapped a ListView item", "Ok");
```

**Comment by samhouts** [comment]
@jsuarezruiz Tests aren't passing. Can you take a look? Thanks!

**Comment by ThaerM** [comment]
Still not working on Xamarin.Forms V4.5.0.617

**Comment by jsuarezruiz** [comment]
@ThaerM We have to complete the review and merge this PR to fix the Collection selection using SwipeView.

File: Xamarin.Forms.Platform.Android/CollectionView/ItemContentView.cs [hunk 4]
```diff
@@ -14,6 +14,11 @@ public ItemContentView(Context context) : base(context)
 		{
 		}
 
+		public void ClickOn()
```
Comment (hartez) [inline Xamarin.Forms.Platform.Android/CollectionView/ItemContentView.cs:17]: Can we make this `internal`?

Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.Android/CollectionView/ItemContentView.cs]: Good point. Yes, changed!.

**Pull request merged**